Add multithreaded Pub/Sub message processing

wuzzeb opened the pull request:
To implement issue #49, add a new multithreaded Pub/Sub message processing feature.
I added extensive haddocks to describe how to use the new feature.
The main benefit over the existing Pub/Sub code (which is left unchanged) is:

- you can make subscription changes at any time
- you can safely recover from networking errors such as the redis servier dying
- you can detect when Redis has actually processed a subscription request and
  handlers will now start receiving messages.

This likely also solves #39 since you can recover from connection failures, and while not directly implementing it, will allow the timeout in #7 since you can manually implement some timeout.

----

The design turned out to be relatively simple:
- There is a single thread which only reads messages from redis
- There is a single thread which only sends subscription changes to redis
- there is a 'PubSubController' which allows any thread to interact with these threads,
  mainly by notifying the sending thread that a subscription change should occur.
- By using two threads and always only reading from one thread and always only sending on another thread.

qrilka commented on src/Database/Redis/PubSub.hs:
```diff
@@ -3,21 +3,39 @@
     publish,
+
+    -- ** Subscribing to channels
+    -- $pubsubexpl
```
some leftover?

wuzzeb replied:
No, I added an explanation of why there are two different pub/sub implementations.  That is a haddock reference to the documentation block at the bottom of the file.

qrilka commented on src/Database/Redis/PubSub.hs:
```diff
@@ -83,6 +101,14 @@ sendCmd cmd       = do
 
+totalPendingChanges :: PubSub -> Int
+totalPendingChanges (PubSub{..}) =
+  updatePending subs $ updatePending unsubs $ updatePending psubs $ updatePending punsubs 0
```
I'd say that sum of pending counts would look more intuitive here but that's just a minor nitpicking...

qrilka replied:
and that also makes me think that `updatePending` is a bit confusing name here as there no actual "update" in it something like `addPending` seems to look better

wuzzeb replied:
Yeah, but I am just reusing the existing updatePending code since I initially was trying on leaving any of the existing pub/sub code unchanged.  Probably better as you say to modify it and then update the old pub/sub code to a new internal function.

qrilka commented on src/Database/Redis/PubSub.hs:
```diff
@@ -212,6 +238,275 @@ pubSub initial callback
                 unless (subCnt == 0 && pending == 0) recv
 
+-- | A Redis channel name
+type RedisChannel = ByteString
```
what about using newtypes to make it more type safe?

wuzzeb replied:
I don't have a big feeling one way or the other, but for the most part I expect the channels to be raw strings with OverloadedStrings enabled so I don't know how much type saftey you would get.  I made them primarily so the haddock documentation looks nice.  But as I said, I don't really lean one way or the other.

qrilka commented on src/Database/Redis/PubSub.hs:
```diff
@@ -212,6 +238,275 @@ pubSub initial callback
+                                 -> [(RedisPChannel, PMessageCallback)] -- ^ the initial pattern subscriptions
+                                 -> m PubSubController
+newPubSubController x y = liftIO $ do
+    cbs <- newTVarIO (fmap (\z -> [z]) $ HM.fromList x)
```
IMO explicit `HM.map` seems to be more readable than `fmap` here

wuzzeb replied:
ok

k-bx commented on src/Database/Redis/Core.hs:
```diff
@@ -2,7 +2,7 @@
 
 module Database.Redis.Core (
-    Connection, connect,
+    Connection(..), connect,
```
I propose keeping this type abstract, while providing `unConnection` deconstructor which would be explicitly said to be internal.

wuzzeb replied:
This is kept internal to hedis, the `Core` module is never exported so this isn't leaking out anywhere in the public API.  I just needed it in PubSub.hs.  Now if you are talking about protecting Core.hs from the rest of hedis itself I guess we can use a `unConnection` destructor.

k-bx replied:
Right, then we don't need this, sorry.

k-bx commented on src/Database/Redis/PubSub.hs:
```diff
@@ -212,6 +238,275 @@ pubSub initial callback
+--
+-- Messages are processed synchronously in the receiving thread, so if the callback
+-- takes a long time it will block other callbacks and other messages from being
+-- received.  If you need to move long-running work to a different thread, I suggest
```
s/I/We/ would be better IMHO

wuzzeb replied:
ok

k-bx commented:
This looks good overall, thank you very much! One last minor nitpick – would be nice to add at least few happy-path tests into the test suite which runs on CI. Otherwise, I'll wait for this and other minor edits and then would be happy to merge!

wuzzeb commented:
Ok, I'll work on adding some tests to the automated test suite plus the small changes tomorrow.

Also, the past few days I have been starting to use this in my own project and found one minor place I would like the API to improve.  Right now you can subscribe to the same channel multiple times and the controller keeps a list of handlers, but you can't unsubscribe an individual handler, you can just unsubscribe in bulk all handlers.  I plan on adding some method to unsubscribe from an individual handler and only one the handler count reaches zero is the channel actually unsubscribed from redis.

k-bx commented:
@wuzzeb all right, sounds reasonable

wuzzeb commented:
- I made the above suggested changes, except I left RedisChannel and RedisPChannel as types instead of newtypes.  I don't think there is much type safety, and it just makes it harder to work with the channels when you are registering them.

- I added a method to unregister only the handlers you registered.  This took me a while to find a nice API and implementation, I think the current one is OK.  It (internally) allocates a unique integer each time you add handlers, and uses that to filter out the handlers on unregister.

- I added some tests to the automatic test suite.  Actually, once I started writing them, I was able to test essentially everything except reconnecting after the redis-server dies.

qrilka commented on src/Database/Redis/PubSub.hs:
```diff
@@ -212,6 +243,343 @@ pubSub initial callback
+        remPChans = filter (\n -> HM.member n pm) pchans
+
+    -- helper functions to filter out handlers that match
+    let filterHandle :: Maybe [(UnregisterHandle,a)] -> Maybe [(UnregisterHandle,a)]
```
what about naming this funciton `removeHandle`? It appears to do a portion of `removeHandles` and doesn't work completely as `filter` so the current name is a bit confusing IMO

k-bx commented:
All right, thanks a lot! Will make a release soon.
